drivers/i3c/i3c_dw.c

```diff
@@ -0,0 +1,2346 @@
+
+		/* Data has been written to the rx fifo by a controller while in target mode*/
+		if (status & INTR_RX_THLD_STAT) {
+			/* while the rx fifo buffer is not empty */
+			while (DATA_BUFFER_STATUS_LEVEL_RX(
+				       sys_read32(config->regs + DATA_BUFFER_STATUS_LEVEL)) > 0) {
+				uint8_t rx_data;
+				/*
+				 * TODO: verify how byte size (lol) data is written to a dword size
+				 * buffer over i3c in target mode
+				 */
+				read_rx_fifo(dev, &rx_data, sizeof(rx_data));
+				/* call function pointer for write */
+				if (target_cb != NULL && target_cb->write_received_cb != NULL) {
+					target_cb->write_received_cb(data->target_config, rx_data);
+				}
+			}
+			sys_write32(INTR_RX_THLD_STAT, config->regs + INTR_STATUS);
+		}
+		/* Read Requested when the CMDQ is empty*/
+		if (status & INTR_READ_REQ_RECV_STAT) {
+			if (target_cb != NULL && target_cb->write_received_cb != NULL) {
+				/* TODO: how should the first byte be handled? */
+				target_cb->read_requested_cb(data->target_config, NULL);
+			}
+			sys_write32(INTR_READ_REQ_RECV_STAT, config->regs + INTR_STATUS);
+		}
+#ifdef CONFIG_I3C_USE_IBI
+		/* IBI TIR request register is addressed and status is updated*/
+		if (status & INTR_IBI_UPDATED_STAT) {
+			k_sem_give(&data->ibi_sts_sem);
+			sys_write32(INTR_IBI_UPDATED_STAT, config->regs + INTR_STATUS);
+		}
+		/* DA has been assigned, could happen after a IBI HJ request */
+		if (status & INTR_DYN_ADDR_ASSGN_STAT) {
+			/* TODO: handle IBI HJ with semaphore */
+			sys_write32(INTR_DYN_ADDR_ASSGN_STAT, config->regs + INTR_STATUS);
+		}
+#endif /* CONFIG_I3C_USE_IBI */
+	}
+
+	return 0;
+}
+
+static void init_scl_timing(const struct device *dev)
+{
+	const struct dw_i3c_config *config = dev->config;
+	struct dw_i3c_data *data = dev->data;
+	uint32_t scl_timing, hcnt, lcnt;
+
+	uint64_t core_rate = config->core_clk;
+
+	/* I3C_OD */
+	hcnt = DIV_ROUND_UP(config->od_thigh_max_ns * core_rate, I3C_PERIOD_NS) - 1;
+	if (hcnt < SCL_I3C_TIMING_CNT_MIN) {
+		hcnt = SCL_I3C_TIMING_CNT_MIN;
+	} else if (hcnt > SCL_I3C_TIMING_CNT_MAX) {
+		hcnt = SCL_I3C_TIMING_CNT_MAX;
+	}
+	lcnt = DIV_ROUND_UP(config->od_tlow_min_ns * core_rate, I3C_PERIOD_NS);
+	if (lcnt < SCL_I3C_TIMING_CNT_MIN) {
+		lcnt = SCL_I3C_TIMING_CNT_MIN;
+	} else if (lcnt > SCL_I3C_TIMING_CNT_MAX) {
+		lcnt = SCL_I3C_TIMING_CNT_MAX;
+	}
+	scl_timing = SCL_I3C_TIMING_HCNT(hcnt) | SCL_I3C_TIMING_LCNT(lcnt);
+	sys_write32(scl_timing, config->regs + SCL_I3C_OD_TIMING);
+
+	/* Set bus free timing to match tlow setting for OD clk config. */
+	sys_write32(BUS_I3C_MST_FREE(lcnt), config->regs + BUS_FREE_TIMING);
+
+	/* I3C_PP */
+	hcnt = DIV_ROUND_UP(I3C_BUS_THIGH_MAX_NS * core_rate, I3C_PERIOD_NS) - 1;
+	if (hcnt < SCL_I3C_TIMING_CNT_MIN) {
+		hcnt = SCL_I3C_TIMING_CNT_MIN;
+	} else if (hcnt > SCL_I3C_TIMING_CNT_MAX) {
+		hcnt = SCL_I3C_TIMING_CNT_MAX;
+	}
+	lcnt = DIV_ROUND_UP(core_rate, I3C_BUS_TYP_I3C_SCL_RATE) - hcnt;
```

Comment:
```suggestion
	lcnt = DIV_ROUND_UP(core_rate, data->common.ctrl_config.scl.i3c) - hcnt;
```

Reply:
How the IP wants to be defined is that it uses the MXDS rate (8, 6, 4, 2 MHz....) and just 'assumes' that everything can get up to 12.5MHz.... but I understand your point with how 'sometimes' a user will want to have a configured rate there